go/pools/refresh_pool.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pools
+
+import (
+	"sync"
+	"time"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+type (
+	// RefreshCheck is a function used to determine if a resource pool should be
+	// refreshed (i.e. closed and reopened)
+	RefreshCheck func() (bool, error)
+
+	// poolRefresh refreshes the pool by calling the RefreshCheck function.
+	// If the RefreshCheck returns true, the pool is closed and reopened.
+	poolRefresh struct {
+		refreshCheck    RefreshCheck
+		refreshInterval time.Duration
+		refreshTicker   *time.Ticker
+		refreshStop     chan struct{}
+		refreshWg       sync.WaitGroup
+
+		pool IRefreshPool
+	}
+)
+
+type IRefreshPool interface {
```

Comment:
I personally don't mind it, but I think the `IName`  is mostly a Java/C# thing, not very idiomatic golang.